### ISSUE TITLE
useBuiltIns: figure out how to remove polyfills that aren't actually used in the codebase

### ISSUE THREAD

**Issue by hzoo** [babel/babel-preset-env#84]
### Don't include if supported environments support it natively #20
- Use preset-env config to determine what parts of core-js to pull in (original PR, #55)

```js
import "babel-polyfill";
```

gets transformed to

```js
import "core-js/modules/es6.map";
import "regenerator-runtime/runtime";
...
```

### Don't include if you have blacklisted it yourself (regenerator for example) #60
- Be able to blacklist transforms or specific builtIns? 

```js
{
  "presets": [
    ["env", {
      "targets": {
        "browsers": ["last 2 versions", "safari >= 7"]
      },
      "blacklist": [
        "transform-regenerator" // won't include "regenerator-runtime/runtime"
      ]
    }]
  ]
}
```

---

### Don't include if the codebase doesn't actually use the builtIn (this)

> this would need to be a whole codebase transformation so it probably won't work in this preset/plugin

The problem with the first heading ^ is that even if you remove the polyfills supported natively in the environment, you might not actually use them in your codebase anyway.

Possible solution: run an analyzer (separate babel pass like [transform-runtime](https://github.com/babel/babel/tree/master/packages/babel-plugin-transform-runtime)) on the whole codebase and remove imports that aren't used anywhere.

EDIT: A big problem atm is that Babel runs per file. So unless we understand the whole codebase it sounds like something webpack could do? cc @TheLarkInn @keyanzhang

**Comment by TheLarkInn** [comment]
So here is an idea (atleast in terms of webpack). //cc @patrickkettner of Modernizr who this could be highly relevant for.

## On demand async polyfill chunks

### Background: ContextModules with webpack 

Since webpack is a static build tool, there is no _real_ dynamic requires, etc. However, there is a static solution to this dynamic nature. 

Given the following builtins are statically located somewhere:

```
--/builtins
  --/somePoly1.js
  --/somePoly2.js
  --/somePoly3.js
```

One can pass an expression to a dependency statement. Specifically `System.import()` or `import()` (for async chunking). 

```javascript 
function asyncLoadPoly(builtInUniqueName) {
  return `import('./builtins' + builtInUniqueName)`
}
```

webpack will create a separate chunk for each possible module that could be loaded from that folder.

This will then return a Promise and one can load that specific poly at runtime: 

```javascript
if (someLogicAboutSpecificJSFeature) {
  asyncLoadPoly('somePoly1.js').then(
    return module.default; 
  ) 
}
```

### Idea

Use this pattern to "dynamically" require x polyfills based on a series of fast logic operations that async load async polyfill chunks as they are needed through out the web application using the above technique. 


### Challenges

* Slow builds: If you are using a crap ton of polys, then you are generating a huge context module with many many asnyc chunks. This could have slower perceived build times. 

### Wins

* Performance (in theory?)
   * you are technically not building as much up front
* Don't have to create builds specifically for each browser

Really this is just one idea that we could leverage, however it could be powerful if we determine the right logic cases up front and what users are targeting.

**Comment by EnoahNetzach** [comment]
@TheLarkInn on demand chunks means that the build step is faster, but runtime it has to be blocking, so more waiting for the client.

@hzoo how can you analyze for method usages (e.g. `foo[bar()].includes()`)?

**Comment by TheLarkInn** [comment]
Well to create context modules it may be slower but I guess the blocking part really is about where the polyfill is needed for which parts of your code.

**Comment by EnoahNetzach** [comment]
So the problem could be reduced to usage detection, then replace the single top-file import of `babel-polyfills` with multiple imports where they are actually used?

I guess this would be possible with ponyfills.

Another solution could be to make the polyfills as lazy chunk loaders so that the first time it is used runtime it gets the actual polyfill and replace itself.
It sounds convoluted, thou.

**Comment by EnoahNetzach** [comment]
Something like 
```es6
featuresToPolyfill.map(({ object, feature }) => {
  object.prototype[feature] = (...args) => {
    object.prototype[feature] = awaitFetchPolyfill(object, feature)
    return object.prototype[feature](...args)
  }
})
```

**Comment by hzoo** [comment]
@EnoahNetzach So for `prototype` methods we assume any usage of it would count as the built-in.

`arr.includes` is using `Array.prototype.includes` but we don't know what `arr` is without type information. We just check the `.includes` part. So it won't have false negatives but it may have false positives?

**Comment by EnoahNetzach** [comment]
@hzoo exactly, that's the problem

**Comment by hzoo** [comment]
Is it really a problem though? My point is that doing that is better than not doing the check at all. It just doesn't save as much if there are false positives

**Comment by EnoahNetzach** [comment]
Oh, I get it.
Well, in this case is far better than having them all always; worst case scenario it doesn't change anything.

But what if someone uses something like this?
```es6
function troll() { return 'includes'; }
obviousArray[troll()]();
```

**Comment by hzoo** [comment]
So that's different lol. It's just a straightforward AST check and we aren't going to support something like that - just add a disclaimer since it's not normal to do that.

**Comment by EnoahNetzach** [comment]
Or maybe leave the option to forcefully import something, if known:
```es6
import `babel-polyfill/a`
import `babel-polyfill/b`

x[f()]()
```

where the programmer knows that `f()` could return either `a` or `b`.

**Comment by DrewML** [comment]
>  we aren't going to support something like that - just add a disclaimer since it's not normal to do that.

We do a similar thing (disclaimer) for [the `Object.assign` transform](https://babeljs.io/docs/plugins/transform-object-assign/)

**Comment by hzoo** [comment]
We can include polyfills in the whitelist/blacklist options if necessary

**Comment by zloirock** [comment]
@TheLarkInn unrealistically.

**Comment by zloirock** [comment]
@hzoo as I wrote, the main problem - we should also parse **all used** dependencies and should know all required features **at the moment of transformation**.

**Comment by hzoo** [comment]
Yeah I don't think we have a way to do that yet standalone unless we do something like make the preset as 2 passes and then write the used polyfills into a tmp directory and read it

**Comment by mikesherov** [comment]
Would it be crazy to consider building out UA specific chunks, and have a special UA that does a UA sniff for which polyfills to load? Kind of like a local polyfill.io?

**Comment by hzoo** [comment]
I don't think so - it seems like a lot of folks are interested?

**Comment by patrickkettner** [comment]
to be clear, I think UA specific builds is dangerous. Feature specific builds (i.e. target es3, es5, etc) is great, but saying here is a chrome-build, here is a IE-build, etc is footguns waiting to be fired

**Comment by mikesherov** [comment]
@patrickkettner not for IE. It's a terminal browser that is no longer being updated with a static set of feature.

**Comment by hzoo** [comment]
I think the main win is targeting non-evergreen browsers (ie 11, safari 9)? Otherwise many would be supporting something like `last 2 browsers` which we can target via `browserslist`

**Comment by Kovensky** [comment]
Runtime loading of polyfills should never be done automatically.

The best case scenario would be to have a generated "polyfill-loader.js" file with an `async` function that does feature tests and dynamic `import`s for the ones that fail; but that would have to be explicitly called by the user, at runtime, with the knowledge that it will return a `Promise`.

It is also **not safe** to execute _any_ javascript until that `Promise` resolves. Any javascript that runs before all polyfills are loaded may be able to capture non-polyfilled objects, may do faulty feature checking, or even load their own shim or polyfill in response.

If your main JS is critical for rendering, you will also get a flash of unrendered content until the polyfills finish loading and it's safe to run JS. At least sync JS doesn't have the potential to show a broken page until it finishes loading (depending on where you place the `script` tag).

**Comment by EnoahNetzach** [comment]
After all I think babel should just have to be able to eliminate dead imports, probably by a simple ast transform per file, then bundlers should have the honor to decide what's really needed when.

It's not a transpiler work to figure out the shape of what the client needs (e.g. chunks), but to give to instruments down the line the tools necessary to make this decision, otherwise I see a duplication of concerns (and compatibility issues).

**Comment by hzoo** [comment]
>  be able to eliminate dead imports, probably by a simple ast transform per file

Not sure how that would work? Currently we would import the polyfill once and the issue was that you might not use that method/polyfill in the whole app (so you can't do a per file check)

**Comment by aickin** [comment]
Hey there, Babel folks!

I've been working on something that basically does this kind of static analysis, and I just made the code public since you all are discussing solving this problem. It's called [`babel-plugin-which-builtins`](https://github.com/aickin/babel-plugin-which-builtins), and it searches the AST for references to new ECMAScript globals, static methods, and instance methods and adds an `import` for the correct polyfill for each.

I think it could be used on individual modules pre-bundling, as the bundler should deal effectively with multiple files that all import the same polyfill. 

Unfortunately, my plugin probably doesn't work well with this project's current strategy of expanding `import 'babel-polyfill'` to a bunch of import declarations that are needed by the env. It would, however, work with a (hypothetical) plugin that did the opposite: remove the polyfill imports that **aren't** needed by the preset. In this theoretical setup, `babel-plugin-which-builtins` would add the polyfill imports needed by the code, and then `babel-preset-env` would remove the imports not needed by the current env. Does that make sense?

It's still very new (just been working on it a week or two), but it's got some tests and I think covers many (most?) basic cases. One case I know that is probably worth covering that is not currently covered is destructuring as property access, like:

```
// not currently covered by babel-plugin-which-builtins, but probably should be.
var { cos } = Math;
```

If y'all want, I'd be happy to offer it under a compatible license. Please let me know what you think, and thanks for all your great work!

**Comment by EnoahNetzach** [comment]
@hzoo the transformer could remove the "global" `babel-polyfill` import and check for method usages that have to be polyfilled, then add the appropriate polyfill `import` in that module.

The so-imported module would just have to pollute the global with polyfills.

**Comment by aickin** [comment]
Sorry for the double post, but I've been thinking about this problem for the last day or so, and I think I don't understand something that was said above:

> this would need to be a whole codebase transformation so it probably won't work in this preset/plugin
> A big problem atm is that Babel runs per file. So unless we understand the whole codebase it sounds like something webpack could do?

I'm probably missing something, but it's not clear to me why this is the case. I think that the output of this preset should be adding the "exactly correct" set of imports, which is the intersection of (builtins used by the code) ∩ (builtins NOT supported natively by env).

It seems to me that as long as the preset adds this exact set of imports, it could be applied at either the file level or at the codebase level. Am I missing something here?

If I'm right, then the preset needs to do a subtraction step to make the intersection. This could be accomplished a few ways:

1. Use `babel-plugin-which-builtins`, which adds the the set of (builtins used by the code), then write a plugin that removes any builtin import supported natively by the env.
1. Use the current `babel-preset-env` with `useBuiltins`, which adds the set of (builtins NOT supported natively by env), then write a plugin that removes any builtin NOT used by the code.
1. Combine the logic of these two plugins so that the intersection is done before the imports are added to the code.

Am I on the right track here? Thanks for you consideration!

**Comment by Kovensky** [comment]
There are a few problems.

1. prototype methods

If you call `.is` on something, does that mean `Object.prototype.is`, or some other method? What about `.then`, or `.find`?

2. order of execution

If your program has more than one module, and the modules don't have the exact same external feature set, then not all polyfills will be installed by the time your code executes.

3. third party modules

If you import an external, third party module that relies on a method that you need to polyfill, the polyfill might get accidentally omitted should the same feature not be also used directly by your own code.

**Comment by aickin** [comment]
@Kovensky thanks for the response! Here are my thoughts:

### Prototype Methods
For prototype methods, I agree that it is hard to know for sure that they are being called, but as @hzoo [said above](https://github.com/babel/babel-preset-env/issues/84#issuecomment-266515856), it's probably best to just always include `Array.prototype.find` every time a `find` property is accessed in the code. This will lead to some false positives (explained more [here](https://github.com/aickin/babel-plugin-which-builtins#instance-methods-may-produce-false-positives)), but it's better than just including all prototype methods. Also, I'm pretty sure that prototype methods are just as much a problem with individual files as they are with a full code base. Do you agree?

### Order of Execution
I'm not clear why order of execution would be a problem, as long `babel-preset-env` adds the correct imports to each individual file. For example, imagine there are two files:

```
// a.js
window.foo = Array.from([1,2,3]);

// b.js
window.bar = new Set([4,5,6]);
```

Ideally `babel-preset-env` would transform these to:

```
// a-post-transform.js
import "core-js/modules/es6.array.from"
window.foo = Array.from([1,2,3]);

// b-post-transform.js
import "core-js/modules/es6.set"
window.bar = new Set([4,5,6]);
```
Each file imports only the polyfills that it needs, and if you bundle them together, the bundler adds the union of the polyfills from all the files. And since each polyfill is imported before it is used, they will all be available by the time they are used. Am I missing something here?

### Third-party Modules
As for third-party modules, I agree that they would be a problem if they use a builtin that your code doesn't use and you don't transform the third-party module with `babel-preset-env`. But this could surely be solved by running `babel-preset-env` on the third-party module file before bundling rather than requiring that `babel-preset-env` only run on the entire codebase, right? It seems to me that it's exactly the same situation as when a third-party module uses ES2015/2016/2017 syntax: you have to then run the third-party module through Babel yourself. Does that make sense?

Thanks for this discussion, and I look forward to all the great things that `babel-preset-env` will be able to do once these features are implemented!

**Comment by zloirock** [comment]
@aickin global polyfills should be added in entry point, before all other js code, otherwise, they could break too many things. For polyfills without global export, you already have it and it's called `babel-runtime`.

> But this could surely be solved by running `babel-preset-env` on the third-party module file before bundling rather than requiring that `babel-preset-env` only run on the entire codebase, right?

Nope. Some libraries have hard polyfills dependencies, for example, `Promise`. But almost all already compiled to ES5. Running `babel` on dependencies is a bad practice and most likely it will break something, makes sense only static analysis for getting a list of required polyfills.

**Comment by aickin** [comment]
@zloirock Thanks for the response!

>  otherwise, they could break too many things

Do you have examples of the problems that could happen or references that I could read to learn more? I'm having a hard time of coming up with obvious problems, but I know you have a ton more experience with the polyfills (having written most of them!).

> But almost all [third-party libraries] already compiled to ES5. Running babel on dependencies is a bad practice

I understand that is true now, but my hope is that `babel-preset-env` will change that. Once we have the ability to ship ES2015/2016/2017 code to browsers, I hope that third-party libraries will start shipping both their ES5 versions and their ES2015/2016/2017 versions, letting the developer decide which features to compile, in both first- and third-party code. There's more to do before this is practical, but `babel-preset-env` gets about 90% of the way towards enabling this idea.

> Running babel on dependencies... most likely it will break something, 

Do you have examples of things that might break or references I could read about this? It seems to me that if running ES2015/2016/2017 syntax transforms on a third-party ES5 codebase breaks something, then that's a bug with the transform that could show up for first-party ES5 code too, right?

> makes sense only static analysis for getting a list of required polyfills.

So, it sounds like your preferred order of operations is:

1. Run `babel-preset-env` on **first-party code only** with `useBuiltins` turned off. This will transform the ES2015/2016/2017 syntax that is not supported by the environment.
1. Bundle first-party and third-party code into one file.
1. Run `babel-preset-env` on the bundled file with all syntax transforms blacklisted and `useBuiltins` turned on.

Is that a fair restatement of what you are arguing for?

Thanks for engaging in this discussion, and for all the great work you do on `core-js`!

**Comment by zloirock** [comment]
> Do you have examples of the problems that could happen or references that I could read to learn more? I'm having a hard time of coming up with obvious problems, but I know you have a ton more experience with the polyfills (having written most of them!).

https://github.com/grigio/meteor-babel/issues/5

> I understand that is true now, but my hope is that `babel-preset-env` will change that.

Unlikely.

> Do you have examples of things that might break or references I could read about this?

https://github.com/zloirock/core-js/issues/189

> So, it sounds like your preferred order of operations is

Something like that.

**Comment by aickin** [comment]
@zloirock Thanks!

> grigio/meteor-babel#5

This confuses me, as this seems like exactly the kind of bug that would have been fixed by running `babel-preset-env` with `useBuiltins` turned on. This problem, as I understand it, was that `Number` was referred to and stored in a data structure before the `Number` constructor polyfill ran. But detecting that earlier reference to `Number` and importing the `Number` constructor polyfill before is the exact problem that `useBuiltins` is supposed to solve!

You may be saying that, aside from this exact example, you just don't trust that any plugin could truly find all the built-in references, and so edge case bugs like this are inevitable. Is that a fair characterization?

> zloirock/core-js#189

Got it. Looks like Babel fixed this particular bug, but I can totally understand that running Babel on a polyfill library is destined to run into some interesting edge cases. FWIW, I agree that third-party libs should be opt-in to ES2015/2016/2017 compilation. 

It is worth noting that whether or not syntax transforms are run on third-party code is orthogonal to whether the `useBuiltins` transform is run on individual files or after bundling.

> Something like that.

Thanks, that helps me understand your position.

**Comment by zloirock** [comment]
@aickin
> This confuses me

We can refer to global variables by different ways. For example, `global.Number`, and we have really many ways of global object detection. It's like `Object['as' + 'sign']`. We can't detect it in this case and in most cases it's not required, but in cases like this, it breaks dependencies.

> Looks like Babel fixed this particular bug

It's just an example. One more - modules transforms adds `"use strict"` and all packages which use top level `this` as a link to the global object will be broken.

**Comment by aickin** [comment]
> We can refer to global variables by different ways... We can't detect it in this case and in most cases it's not required, but in cases like this, it breaks dependencies.

So does it follow that you're against the `useBuiltins` feature being discussed in this bug, since it is too hard to figure out when globals are being used?

> It's just an example.

I understand and agree on this. ES2015/2016/2017 compilation should be opt in.

**Comment by zloirock** [comment]
@aickin
> So does it follow that you're against the `useBuiltins` feature being discussed in this bug, since it is too hard to figure out when globals are being used?

Nope. I just mean that polyfill should be before all other code. We can miss some polyfills, it's not a very big problem, but adding polyfills after another code can be dangerous.

**Comment by laszbalo** [comment]
Can we just define manually what features we need, e.g. with a includeIfMissing object literal, and let env compute the intersection?

**Comment by romulof** [comment]
Is it not possible to use `babel-plugin-transform-runtime` together with `babel-preset-env`?

**Comment by kaisermann** [comment]
I'm also looking for the same  functionality as @romulof.

**Comment by romulof** [comment]
I ended up disabling `useBuiltIns` and reverting to `babel-plugin-transform-runtime`.
My bundle even shrunk a few kilobytes.

**Comment by kaisermann** [comment]
@romulof Would you care to explain how you did it? Thanks!

**Comment by romulof** [comment]
I leave all my Babel config inside webpack config, and all build parameters are processed by env vars using https://github.com/af/envalid.

Here's it's config object. I think it's pretty self-explanatory:
```es6
      {
        cacheDirectory: env.isDev,
        presets: [
          'react',
          ['env', {
            debug: true,
            modules: false, // Using webpack 2
            targets: {
              // browserlist var is shared with autoprefixer
              browsers: browserlist,
            },
            exclude: [
              // Leave regenerator to transform-runtime
              'transform-regenerator',
            ], 
          }],
        ],
        plugins: compact([
          'lodash',
          'transform-decorators-legacy',
          'transform-class-properties',
          'transform-export-extensions',
          'transform-object-rest-spread',
          env.isProduction && 'transform-react-remove-prop-types',
          env.isProduction && 'transform-react-constant-elements',
          env.isProduction && 'transform-react-inline-elements',
          env.isProduction && 'transform-dev-warning',
          ['transform-runtime', {
            // Don't use regenerator in dev builds. Hard to debug.
            regenerator: env.isProduction,
          }],
          env.isDev && 'react-hot-loader/babel',
          env.isDev && 'transform-react-jsx-self',
          env.isDev && 'transform-react-jsx-source',
        ]),
      }
```

Right now this is working for me, but might not apply to all cases.

**Comment by yavorsky** [comment]
@romulof Nice! Why are you not using `env` field for babel configuration?https://babeljs.io/docs/usage/babelrc/#env-option
```js
{
"plugins": ["transform-decorators-legacy", "other common plugins"], // common plugins
"env": {
    "production": {
      "plugins": ["regenerator"] // plugins for prod
    },
    "development": {
      "plugins": ["react-hot-loader/babel"] // plugins for dev
    }
  }
}
```

**Comment by romulof** [comment]
It's just to make the whole Webpack config uniform. In the end it works the same way.

PS: `compact` is from Lodash. It clears `undefined`/`null` entries from arrays.

**Comment by kaisermann** [comment]
Ok, got it working. But `transform-runtime` STILL does not polyfill only what is listed by `babel-preset-env`. I mean, it keeps, for example, polyfilling `Object.assign()` even if I'm targeting `chrome: 52`. Perhaps I understood something wrong?

My babel configuration:

```js
{
  "presets": [
    ["env", {
      "modules": false,
      "loose": true,
      "debug": true,
      "targets": {
        "browsers": "chrome 52",
      }
    }]
  ],
  "plugins": [
    ["transform-runtime", {
      "helpers": false,
      "polyfill": true,
      "regenerator": true,
      "moduleName": "babel-runtime"
    }],
    ["external-helpers"]
  ]
}
```

**Comment by romulof** [comment]
`transform-runtime` is a separate Babel plugin. It transforms ES6/7/next calls to `babel-polyfill` libraries. You can disable some of this features using it's own configuration.

**Comment by hzoo** [comment]
Ok I think i'm going to try what @aickin was talking about. It's actually just what [transform-runtime](https://github.com/babel/babel/tree/master/packages/babel-plugin-transform-runtime) already does, except we are now going to add the polyfill instead of rewriting to reference babel-runtime. So we just add the used imports and then remove the ones provided by the env. Will post an example later, still wip

**Comment by hzoo** [comment]
Ok I made https://github.com/babel/babel-preset-env/pull/241

only issue not covered in https://github.com/babel/babel-preset-env/issues/84#issuecomment-267235999 is regarding the rest of node_modules (3rd party). Workaround is a disclaimer of the option and to use `includes` instead.

**Comment by hzoo** [comment]
Merged ^, testing in v2.0.0-alpha.5

**Comment by hzoo** [comment]
Closing in favor of a new issue I made: https://github.com/babel/babel-preset-env/issues/284

since we kinda got this working but need some more work for 3rd party modules
